src/mds/Locker.cc

```diff
@@ -4786,12 +4831,70 @@ void Locker::scatter_tick()
 	       << *lock << " " << *lock->get_parent() << dendl;
       continue;
     }
-    if (now - lock->get_update_stamp() < g_conf()->mds_scatter_nudge_interval)
+
+    if (!mdr && now - lock->get_update_stamp() < g_conf()->mds_scatter_nudge_interval)
       break;
     updated_scatterlocks.pop_front();
-    scatter_nudge(lock, 0);
+
+    if (lock->get_type() == CEPH_LOCK_INEST)
+      dout(20) << __func__ << " lock: " << *lock << " of " << *lock->get_parent()
+	<< ", update_stamp: " << lock->get_update_stamp() << ", lock's parent's rstat_dirty_from:"
+	<< static_cast<CInode*>(lock->get_parent())->get_rstat_dirty_from() << ", lock's parent's subdir rstat_dirty_from:"
+	<< static_cast<CInode*>(lock->get_parent())->get_dirfrags_rstat_dirty_from()
+	<< ", mdr's rstats_propagate_time: " << (mdr ? mdr->get_op_stamp() : utime_t()) << dendl;
+
+    if (mdr) {
+
+      if (lock->is_nudged_by(mdr->reqid)) {
+	updated_scatterlocks.push_back(lock->get_updated_item());
+	nudged++;
+	// the lock is nudged by myself previously, and there must
+	// be a C_Locker_RstatFlush that will be triggered, so just
+	// skip this lock for now.
+	continue;
+      }
+
+      CInode* ino = static_cast<CInode*>(lock->get_parent());
+
+      if (lock->get_type() == CEPH_LOCK_INEST
+	  && ino->need_to_nudge(mdr->in[0], mdr->get_op_stamp())) {
+        dout(20) << __func__ << " " << (mdr ? mdr->reqid : metareqid_t())
+          << " need to nudge scatterlock: " << *lock << " of " << *lock->get_parent() << dendl;
+
+        scatter_nudge(lock, new C_Locker_RstatFlush(this, mdr, lock));
```

Comment:
still allocate a C_Locker_RstatFlush for each lock?  I think it's better to use a MDSGatherBuilder and only allocate one C_Locker_RstatFlush. In my option,  CInode::need_to_nudge is not a cheap function

Reply:
@ukernel Thanks for your review:-) I'm trying to modify the code as you suggested and make CInode::need_to_nudge cheap:-)